### PR TITLE
Holmes/metadata

### PR DESCRIPTION
**Pull request by holmes**
I'm not entirely sure that this is appropriate in this library, but it does seem like it would be a good central spot for this kind of information.

### CODE REVIEW

**Comment by holmes** [comment]
I thought of that approach, but the problem is how to handle unknown or custom keys. In this approach you still get the values that "should" be there but also have access to the raw map.

I'll see if I can figure out how to convert the raw metadata map to a custom type.

File: src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            returnValue.add(urlMap.get("url"));
+        }
+
+        return returnValue.toArray(new String[returnValue.size()]);
+    }
+
+    /**
+     * Metadata for an undefined media type.
+     */
+    public static final class GenericMetaData extends MediaMetadata {
```
Comment (holmes) [inline src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java]: We create concrete classes for each type.

File: src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    public static final String SERIES_TITLE = "seriesTitle";
+    public static final String STUDIO = "studio";
+    public static final String SUBTITLE = "subtitle";
+    public static final String TITLE = "title";
+    public static final String TRACK_NUMBER = "trackNumber";
+    public static final String WIDTH = "width";
+
+    @JsonProperty
+    @JsonSerialize(include = JsonSerialize.Inclusion.NON_NULL)
+    public Map<String, Object> metadata = new HashMap<String, Object>();
```
Comment (holmes) [inline src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java]: The abstract base class has the map.

File: src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+    @JsonAnySetter
+    public void unknownSetter(String key, Object value) {
```
Comment (holmes) [inline src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java]: And the map is populated via this.

File: src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+import org.codehaus.jackson.annotate.JsonTypeInfo;
+import org.codehaus.jackson.map.annotate.JsonSerialize;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "metadataType")
+@JsonSubTypes({
```
Comment (holmes) [inline src/main/java/su/litvak/chromecast/api/v2/MediaMetadata.java]: I was able to figure out how to create instances of the right type via the `metadataType` property, but I can't get the instances to be populated with a Map. So using `JsonAnySetter` as a workaround.

**Comment by vitalidze** [comment]
I like this approach much more. However I have several comments:

1) If possible, known fields must be defined as fields in appropriate classes.
2) All unknown fields must be put into Map
3) I am not sure how to handle "Date" fields correctly, there are several ways a date can be represented in JSON.
4) There must be a fallback type of "MetaData", which will be used in case there are no "metadataType" field, or it is not expected.
5) If possible, hide `@JsonAnySetter`, i.e. make access `private` or `protected`.
6) Please write unit tests for all types of metadata.

I want this to suite all other library classes design as much as possible. Also it makes sense to make `MetaData` objects immutable, or at least most fields immutable.

Let me know what do you think.

**Comment by holmes** [comment]
Sounds fair. 

One big question about dates:
I didn't want to force a date library on the end user, nor did I want to roll my own. Hence my keeping them as Strings. The docs say they're ISO-8601. Thoughts on what you want to do there?

Everything else you mentioned sounds good.

**Comment by holmes** [comment]
FWIW, Pandora doesn't even send dates:

```
      "media": {
        "metadata": {
          "metadataType": 3,
          "albumName": "Definitive Greatest Hits",
          "title": "You're Gonna Miss This",
          "albumArtist": "Trace Adkins",
          "artist": "Trace Adkins",
          "images": [
            {
              "url": "http://mediaserver-cont-sv5-2-v4v6.pandora.com/images/public/gracenote/albumart/5/5/7/3/05099964883755_500W_500H.jpg"
            }
          ],
          "releaseDate": ""
        },
```

**Comment by vitalidze** [comment]
> I didn't want to force a date library on the end user, nor did I want to roll my own. Hence my keeping them as Strings. The docs say they're ISO-8601. Thoughts on what you want to do there?

I couldn't find any examples myself either. I think it makes sense to store them as strings.

**Comment by holmes** [comment]
So after implementing this, I'm not convinced this is a good API.

The use case I'm looking at is to get a set of known keys, and update the status from them. It doesn't matter what type it is. The ones that have values should copy them over, the ones w/o values should be null'd out. 

With this API I'd have to switch on the type, get the values from the type and then null out the other values. I think it'd be easier to have a set of keys and loop over them to get the appropriate values.

I can see cases where an API might want to find the type and extract certain keys from them, but then they could just use the predefined keys to extract data from the `Map`.

Thoughts?

**Comment by vitalidze** [comment]
Well, then maybe it makes sense to leave it as is. I mean just have a `Map`. And maybe define some set of known keys as `public static final String` fields somewhere in `Media` class.

**Comment by holmes** [comment]
Simple keys it is. I'm debating on adding an enum for MetadataType. I'll attach as a separate commit so you can keep/remove as you want.

**Pull request merged**